tests/v3/connection_state_test.c

```diff
@@ -3682,3 +3685,146 @@ AWS_TEST_CASE_FIXTURE(
     s_test_mqtt_connection_ping_double_scenario_fn,
     s_clean_up_mqtt_server_fn,
     &test_data)
+
+static bool s_is_termination_completed(void *arg) {
+    struct mqtt_connection_state_test *state_test_data = arg;
+    return state_test_data->connection_terminated;
+}
+
+static void s_wait_for_termination_to_complete(struct mqtt_connection_state_test *state_test_data) {
+    aws_mutex_lock(&state_test_data->lock);
+    aws_condition_variable_wait_pred(
+        &state_test_data->cvar, &state_test_data->lock, s_is_termination_completed, state_test_data);
+    state_test_data->connection_terminated = false;
+    aws_mutex_unlock(&state_test_data->lock);
+}
+
+static void s_on_connection_termination_fn(void *userdata) {
+    struct mqtt_connection_state_test *state_test_data = (struct mqtt_connection_state_test *)userdata;
+
+    aws_mutex_lock(&state_test_data->lock);
+    state_test_data->connection_termination_calls += 1;
+    aws_mutex_unlock(&state_test_data->lock);
+
+    state_test_data->connection_terminated = true;
+}
+
+static int s_clean_up_mqtt_server_with_termination_fn(struct aws_allocator *allocator, int setup_result, void *ctx) {
```

Comment:
Could we just add the termination change directly to `s_clean_up_mqtt_server_fn`? In that case we could have termination callback attached to all unit tests.

Reply:
Initially I did exactly that. But then I thought that separating tested functionality as much as possible would be better. Otherwise, if for some reason `on_termination` callback breaks, it'll cause all other tests in CI to fail, making it more difficult to find the cause of the issue. With the current approach we'll see a couple `test_on_termination_callback_*` failed, and it'll be easy enough to understand where to look.
Saying that, I see that enabling `on_termination` callback for all tests may cover additional cases that I missed or that will appear in the future.

Reply:
I think we can have both. A separate simply for `on_termination` and also applied it to all other tests. If the separate one failed, that's good enough for us to understand that the other tests are also failing because of the `on_termination`.

Reply:
That sounds reasonable, I'll enable on_termination for all tests